### PR TITLE
Session middleware test

### PR DESCRIPTION
**Pull request by kamek-pf**
Test case for #392

### CODE REVIEW

**Comment by kamek-pf** [comment]
So this test won't pass. Swap lines 1008 and 1009 and it does.

**Comment by DoumanAsh** [comment]
Thanks I have difficulties to write these tests because I cannot use `session` feature on windows as of now.
I'll take a look at your test later on

File: tests/test_middleware.rs [hunk 1]
```diff
@@ -993,3 +993,74 @@ fn test_resource_middleware_async_chain_with_error() {
...
+    });
+
+    let request_set = srv.get().uri(srv.url("/first")).finish().unwrap();
+    let response_set = srv.execute(request_set.send()).unwrap();
+
+    // we grab the "set-cookie" header from the response and use it as our "cookie" header for the next request
+    let cookie: &str = response_set.headers()["set-cookie"]
+        .to_str()
+        .unwrap()
+        .split(";") // cut off the last part ("; HttpOnly; Path=/"")
```
Comment (DoumanAsh) [inline tests/test_middleware.rs]: Should it be rsplit?

**Comment by fafhrd91** [comment]
@DoumanAsh could you review and decide what to do with it

**Comment by DoumanAsh** [comment]
@kamek-pf Currently test doesn't pass on https://github.com/actix/actix-web/pull/395/files#diff-48d46aa5fa3906ed3a97ad3e4e6ab7dcR1034
I don't see what is wrong with it and why session is not restored, so I'd need to review Session middleware code before to have proper comments

And my PR has the same issue 😄 
I wonder if we should either stick to your PR or mine #393
It is impossible, at the moment, for me to build with `session` feature due to `ring` issues so I wouldn't mind to close it up and help you to investigate issue

**Comment by kamek-pf** [comment]
Oh sorry, I should have made this more clear :
It appears that if a single value contains problematic characters, then the session middleware is unable to restore anything at all.

The behavior you're seeing here is consistent with my observations.

**Comment by DoumanAsh** [comment]
Ah thanks, that clarifies it as I was wondering what was wrong.

Then given than we both have the same PR, would you like me to continue work on issue in my PR and close your PR? Or vice versa?

**Comment by kamek-pf** [comment]
They do look identical now, I don't mind at all which one gets merged, my goal was just to highlight the issue :p

**Comment by DoumanAsh** [comment]
Ok, I'll close it in favor of my PR #393. hopefully I'll get to bottom of this soon enough

**Pull request closed without merging**